design/als-design.md

```diff
@@ -0,0 +1,68 @@
+# AccessLog Service (ALS) Support
+
+Status: Draft
+
+## Abstract
+
+This document proposed a design for supporting Envoy's AccessLog Service capabilities in Contour.
+
+## Background
+
+Envoy has the ability to send rich access logs to access logging sinks.
+With AccessLog Service (ALS) enabled, Envoy can send access log messages using gRPC streams to a gRPC access log service. 
+
+This document focuses on configuring access log sinks and enabling Envoy to send access log messages via Contour.
+
+## Goals
+- Enable users to specify ALS configuration
+- Enable users to send HTTP access logs for specific listeners 
+
+## Non Goals
+- TCP Access Log Support
+
+## High-Level Design
+Contour will add support for Envoy's HTTP Access Log Service.
+ALS information will be specified as a static cluster for Envoy and Contour controller will set `access_log` field for each HTTPConnectionManager filter.
+
+## Detailed Design
+To enable Envoy to send access logs to a sink; first a static cluster needs to be set for Envoy.
+There are two ways to add this configuration:
+- ExtensionService
+- Contour Configuration file
+
+### ExtensionService
+Contour administrator/operator can add an ExtensionService with ALS configuration details.
+As part of `contour bootstrap`; the ExtensionService used to configure a static ALS cluster for Envoy.
+Contour controller will set `access_log` field for each HTTP filter.
+
+An annotation `projectcontour.io/als: true` needs to be added to the particular ExtensionService.
```

Comment:
This annotation pattern is an interesting one and not one we have used already in Contour

It has the downside of making `boostrap` and `serve` look through all `ExtensionService` resources on startup, but makes it so we don't have to pass the same config in two places since we need to tell both `serve` and `bootstrap` about the ALS service.

Reply:
Yeh, the idea of using an annotation is OK in theory, but in practice I suggest that the name of the annotation should be much more specific (ie. it obviously means "use this as a default log sink"), and it somehow needs to be tied to a specific envoy deployment (there can be more than one in the cluster).

Perhaps this use case can be configured more directly in the Contour flags, or in the operator?

Reply:
The point about multiple Envoy deployments is a very good one, perhaps the annotation could be something like:
`projectcontour.io/envoy-log-sink: <string>" where string is a value that's defaulted, but can be tuned individually on the `contour serve` and `contour bootsrap` commands?

The whole problem is that the bootstrap and the contour deployment need to have the same info configured somehow, so we either need to make a way for them to coordinate, or configure things twice, which seems very prone to failure.

Reply:
👍  to leaving the door open for multiple ALS in the future.

Reply:
Yes, I would suggest we add 2 annotations that would be specific to ALS
`projectcontour.io/als-admitted: true` -> Indicates whether to configure this ALS
`projectcontour.io/envoy-log-sink: <name>` -> ALS name or default.